### ISSUE TITLE
InlineLauncher: adding PhaseBarriers, Reservations, etc.

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#90]
The current InlineLauncher API does not include the ability to add PhaseBarriers, Reservations, etc. Should it? I believe that `add_wait_barrier` would make sense at the least, and Sean has argued for a semantics of inlines that allows the others to make sense as well (by interpreting the scope from inline to explicit unmap as an implicit task scope).

**Comment by lightsighter** [comment]
I intentionally omitted those things from the InlineLauncher.  If a parent task wants to do those things, it can feel free to wait itself.  There are benefits to making the runtime aware of those things for other operations which are all going to get deferred.  But doing an inline mapping is inherently a blocking operation so there is no benefit to composing it with anything.  

If you argue that map/unmap is going to be an implicit task scope, all notion of a serial program semantics goes out the window.  If any interfering operation is launched in between those two things, what is the program order?   This also has the implication that the runtime can no longer do the automatic unmap/remap trick to avoid deadlock because there is no definition of program order.  The exception to this is if you guarantee that there are no interfering operations launched between the map and unmap call, but that feels overly restrictive.  This would be very easy for me to implement, but I'm guessing it would make people unhappy.

The way I see it, inline mappings are productivity feature and not a performance feature.  Therefore they should have very simple semantics and should never be used in performance critical sections of code.

**Comment by streichler** [comment]
Inline mappings do not have blocking semantics.  Forcing somebody to explicitly wait on a barrier before launching an inline mapping is worse than having it delay when their eventual call to wait_until_valid() returns.

And I believe that serial execution semantics still work as long as explicit inline map/unmap calls are properly nested.  The implicit unmap/remap that occurs around a task launch inside an inline mapping is similar to the way that a parent task temporarily cedes ownership of a region to a subtask with the right privileges.

Is there significant work to add it, or are these fields available due to inheritance already?

**Comment by streichler** [comment]
All issues need owners...

**Comment by lightsighter** [comment]
Fixed with commit db71805

**Comment by elliottslaughter** [comment]
@lightsighter Just so we're clear, what are the semantics? E.g. when would an arrive barrier trigger?

**Comment by lightsighter** [comment]
An arrive barrier triggers whenever the region is unmapped, either explicitly by the application or implicitly by the runtime when it gets unmapped/remapping around a task launch or some other operation.
